### ISSUE TITLE
Subproc interferes with startup, causes timeout

### ISSUE THREAD

**Issue by scopatz** [ipython/ipykernel#94]
I am writing with regards to an issue that was first reported for Calysto/xonsh_kernel#2, but is something that we are also seeing with the mainline [xonsh kernel](https://github.com/scopatz/xonsh/blob/master/xonsh/jupyter_kernel.py) as well.

On Jupyter v 4.0.6 and ipykernel v4.1.1 on at least Linux, we consistently see timeout issues, like the following when starting new xonsh kernels.

```console
scopatz@localhost ~ $ jupyter notebook --no-browser
[I 15:47:39.302 NotebookApp] Serving notebooks from local directory: /home/scopatz
[I 15:47:39.302 NotebookApp] 0 active kernels 
[I 15:47:39.302 NotebookApp] The IPython Notebook is running at: http://localhost:8888/
[I 15:47:39.302 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[I 15:47:46.134 NotebookApp] Creating new notebook in 
[I 15:47:47.003 NotebookApp] Kernel started: 9e3c7307-c9e6-4896-bf5d-b5bc302608ea
[W 15:47:57.112 NotebookApp] Timeout waiting for kernel_info reply from 9e3c7307-c9e6-4896-bf5d-b5bc302608ea
```

This only seems to happen when xonsh is sourcing configurations from foreign shells (it loads Bash by default).  This involves running a [subprocess and waiting for some output](https://github.com/scopatz/xonsh/blob/master/xonsh/foreign_shells.py#L141). 

Once the timeout fails, no input from the notebook interface is accepted and the subproc stays alive, eats up 100% CPU, and must be manually SIGKILL'd.  

My guess, though I am not an expert on the Jupyter / IPython internals, is that this interferes with the ability to issue a heartbeat. 

My questions are:

* Is this really a heartbeat issue?
* If so, is there a way to make ipykernel robust in the face of subprocess?
* or do we need to run the subprocess on another thread so as to not block the heartbeat?
* If it isn't a heartbeat, what is going on here? 

It is worth noting that back in late Sept. and early Oct. 2015 this *was* working without significant changes in the xonsh startup procedure. I am reporting here because I believe that something about IPython, Jupyter, or PyZMQ has changed in this time.

Any help would be greatly appreciated! Thanks to @lmmx, @blink1073, and @IanSudbery for help in identifying this issue.

**Comment by takluyver** [comment]
I don't think it's likely to be a heartbeat issue - I'm not sure that we even use heartbeats in the default case, and if you're reusing our kernel machinery, the heartbeat should be started in a separate thread, so running subprocesses in the main thread shouldn't block it.

However, when we start a kernel up, we send it a kernel_info_request, both as a way of getting information about it, and so that the reply acts as a "kernel ready" signal. I don't know what we do when that times out, but I'd guess it's a less well tested corner of the code.

**Comment by minrk** [comment]
As @takluyver explained, it's not a heartbeat issue. It's a failure to reply to a first kernel_info_request. If that takes a long time, the first reply will not be sent, but this shouldn't be a major problem, because kernels can be slow to start up. The UI *should* be fine when this happens. The main things the reply is used for are:

- setting the initial busy/idle state (initial state is busy until an idle message is received)
- setting the highlighting language, if it's not already in the notebook metadata

If the reply is never sent but the kernel starts up otherwise correctly, the worst consequence *should* be that you have to wait ten seconds before executing any code, and the highlighting might be wrong.

Perhaps there's something in the bash config that's causing it to do extra work, such as a prompt for input happening for some reason?. I've grabbed master of everything, and it's starting up without any delay. If I add an artificial delay (`sleep 15` in DEFAULT_BASH_FUNCSCMD), then I see the expected Timeout, but once the command returns, the kernel is responsive and well behaved.

Are there any more details about the systems that can reproduce this issue, since it doesn't appear to be general?

**Comment by scopatz** [comment]
Thanks both, this has been closed by scopatz/xonsh#625
